docs/proposals/disk-accounting.md

```diff
@@ -0,0 +1,641 @@
+
+**Pros**:
+
+* This is the least-intrusive approach.
+
+* It will work off the box without requiring any additional configuration.
+
+**Cons**:
+
+* `du` can consume a lot of cpu and memory. There have been several issues reported against the kubelet in the past that were related to `du`.
+
+* It is time consuming. Cannot be run frequently. Requires special handling to constrain resource usage - setting lower nice value or running in a sub-container.
+
+* Can block container deletion by keeping file descriptors open.
+
+
+2. Linux gid based Disk Quota
+
+	[Disk quota](https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Storage_Administration_Guide/ch-disk-quotas.html) feature provided by the linux kernel can be used to track the usage of image layers. Ideally, we need `project` support for disk quota, which let’s us track usage of directory hierarchies using `project ids`. Unfortunately, that feature is only available for zfs filesystems. Since most of our distributions use `ext4` by default, we will have to use either `uid` or `gid` based quota tracking.
+
+	Both `uids` and `gids` are meant for security. Overloading that concept for disk tracking is painful and ugly. But, that is what we have today.
+
+	Kubelet needs to define a gid for tracking image layers and make that gid or group the owner of `/var/lib/docker/[aufs | overlayfs]` recursively. Once this is done, the quota sub-system in the kernel will report the blocks being consumed by the storage driver on the underlying partition.
+
+	Since this number also includes the container’s writable layer, we will have to somehow subtract that usage from the overall usage of the storage driver directory. Luckily, we can use the same mechanism for tracking container’s writable layer. Once we apply a different `gid` to the container’s writable layer, which is located under `/var/lib/docker/<storage_driver>/diff/<container_id>`, the quota subsystem will not include the container’s writable layer usage.
+
+**Pros**:
+
+* Low overhead tracking provided by the kernel.
+
+
+**Cons**
+
+* Requires updates to default ownership on docker’s internal storage driver directories. We will have to deal with storage driver implementation implementation details in any approach that is not docker native.
+
+* Requires additional node configuration - quota subsystem needs to be setup on the node. This can either be automated or made a requirement for the node.
+
+* Kubelet needs to perform gid management. A range of gids have to allocated to the kubelet for the purposes of quota management. This range must not be used for any other purposes out of band.
+
+* Breaks `docker save` semantics. Since kubernetes assumes immutable images, this is not a blocker. To support quota in docker, we will need user-namespaces along with custom gid mapping for each container. This feature does not exist today.
```

Comment:
Allowing such a functionality will essentially eliminate use of quota for writable layer tracking. I can see why it is useful though. It could be a cluster configuration option - Need better accounting vs I care only about CI. 

@derekwaynecarr: Is it possible to run a side-car that waits for the main container to exit, identifies the main container's ID through Kube APIs (its there in container Status), and then perform a commit and a push?